### PR TITLE
feat(PieChart): semiCircle + legendShowValues

### PR DESCRIPTION
**Pull request by murdore**
Adds optional semiCircle (half-donut) mode and legendShowValues (right-aligned value column in legend). Backward-compatible. check + lint + build pass. hold-and-fold.

### CODE REVIEW

File: src/lib/PieChart/PieChart.svelte [hunk 1]
```diff
@@ -123,11 +183,29 @@
   {#if isEmpty && typeof empty === 'function'}
     <div class="chart-empty">{@render empty()}</div>
   {:else}
-    {#if showLegend}
+    {#if showLegend && !legendShowValues}
       <Legend items={legendItems} position="top" />
     {/if}
```
Comment (Tara-ag) [inline src/lib/PieChart/PieChart.svelte:200]: 💡 MINOR: The new legend rows rendered when `legendShowValues=true` use `<div>` elements without ARIA roles or accessible markup. Consider adding `role="list"` to the container and `role="listitem"` to each row for screen reader compatibility, or use semantic HTML like `<ul>`/`<li>`.

```svelte
<!-- Instead of: -->
<div class="pie-legend-values">
  <div class="pie-legend-row">...</div>
</div>

<!-- Consider: -->
<ul class="pie-legend-values">
  <li class="pie-legend-row">...</li>
</ul>
```

File: src/lib/PieChart/properties.ts [hunk 2]
```diff
@@ -29,6 +29,8 @@ export type OptionalPieChartProperties = {
   empty?: Snippet;
   testId?: string;
   classes?: string;
+  semiCircle?: boolean;
```
Comment (Tara-ag) [inline src/lib/PieChart/properties.ts:32]: 💬 SUGGESTION: The new props `semiCircle` and `legendShowValues` are correctly typed as optional booleans with defaults in the component. Ensure the documentation at `docs/PieChart.md` is updated to include these new props and their corresponding CSS variables (`--piechart-semi-aspect-ratio`, `--piechart-legend-gap`, `--piechart-legend-padding`, `--piechart-legend-value-font-size`, `--piechart-legend-value-color`).

File: src/lib/PieChart/PieChart.svelte [hunk 3]
```diff
@@ -40,39 +43,96 @@
   let mouseX = $state(0);
   let mouseY = $state(0);
 
+  // Aspect ratio read from --piechart-semi-aspect-ratio CSS variable.
+  // Updated on mount and whenever the container element changes.
+  let semiAspectRatioCssVar = $state(2);
+
+  onMount(() => {
+    if (containerEl !== null) {
```
Comment (Tara-ag) [inline src/lib/PieChart/PieChart.svelte]: 💡 MINOR: The `onMount` hook reads `--piechart-semi-aspect-ratio` once at mount time. If the consumer dynamically changes this CSS variable (e.g., via media queries or theme switching), the component won't reflect the new value until remount. Consider using a `ResizeObserver` on the container or a CSS custom property observer if dynamic updates are a use case.

**Review by Tara-ag — APPROVED** [review]
## Review Summary

**Files reviewed:** 2 (`PieChart.svelte`, `properties.ts`)

**New issues raised:**
- 💡 MINOR: 2 (accessibility of legend list structure, CSS variable reactivity)
- 💬 SUGGESTION: 1 (documentation update needed)

**Assessment:**

✅ **Backward compatible** — New `semiCircle` and `legendShowValues` props are optional with sensible defaults (`false`), preserving existing behavior.

✅ **Public API integrity** — No breaking changes to props, events, or exports. The `aspectRatio` prop behavior is refined (now `undefined` by default with fallback logic) but remains backward compatible.

✅ **SSR safe** — `getComputedStyle` access is properly guarded inside `onMount`.

✅ **Svelte 5 idiomatic** — Correct use of `$derived`, `$derived.by`, and `$state` for the new reactive computations.

**Notes:**
- The new CSS custom properties (`--piechart-semi-aspect-ratio`, `--piechart-legend-*`) should be documented in `docs/PieChart.md`
- Consider using semantic `<ul>`/`<li>` for the legend values list for better accessibility

**Recommendation:** Approve with minor suggestions.

File: src/lib/PieChart/PieChart.svelte [hunk 4]
```diff
@@ -40,39 +43,105 @@
   let mouseX = $state(0);
   let mouseY = $state(0);
 
+  // Aspect ratio read from --piechart-semi-aspect-ratio CSS variable.
+  // Uses $effect so it re-reads whenever containerEl binds or semiCircle changes
+  // (e.g. media-query theme switch), not just on initial mount.
+  let semiAspectRatioCssVar = $state(2);
+
+  // eslint-disable-next-line no-restricted-syntax
+  $effect(() => {
```
Comment (Tara-ag) [inline src/lib/PieChart/PieChart.svelte:52]: 💬 SUGGESTION: The `$effect` that reads `--piechart-semi-aspect-ratio` CSS variable tracks `semiCircle` via `void semiCircle` but does not track `containerEl` binding changes. If the container element is recreated (e.g., via `{#key}` block), the effect won't re-run until `semiCircle` changes. Consider also tracking `containerEl`:

```svelte
$effect(() => {
  void semiCircle;
  void containerEl; // Re-run when container element changes
  if (containerEl !== null) {
    // ... existing logic
  }
});
```

File: src/lib/_chart/format.ts
Comment (Tara-ag) [inline src/lib/_chart/format.ts:1]: 💡 MINOR: The new `formatNumberIndian` utility function is being exported as part of the public API but lacks unit tests. Consider adding tests to cover:
- Values at threshold boundaries (1e7, 1e5, 1e3)
- Negative values
- Zero handling
- Values between thresholds

```typescript
// Example test cases:
expect(formatNumberIndian(15000000)).toBe('1.5Cr');
expect(formatNumberIndian(250000)).toBe('2.5L');
expect(formatNumberIndian(5000)).toBe('5K');
expect(formatNumberIndian(-15000000)).toBe('-1.5Cr');
```

File: src/lib/_chart/ChartContainer.svelte [hunk 5]
```diff
@@ -14,19 +14,27 @@
 
   let containerEl: HTMLDivElement | null = $state(null);
 
-  onMount(() => {
+  function measure() {
     if (containerEl === null) {
       return;
     }
+    const rect = containerEl.getBoundingClientRect();
+    const w = Math.round(rect.width);
```
Comment (Tara-ag) [inline src/lib/_chart/ChartContainer.svelte:23]: ⚠️ MAJOR: The new `$effect` for measuring on `aspectRatio` changes introduces a potential memory leak. The `ResizeObserver` from `onMount` is not cleaned up when the component is destroyed - only when `onMount`'s cleanup function runs. However, the `$effect` runs independently and calls `measure()` which accesses `containerEl`. If the component unmounts while the effect is scheduled, `containerEl` could be null or stale.

Consider ensuring proper cleanup synchronization between the `$effect` and `onMount`:

```svelte
let cleanupFns: Array<() => void> = [];

$effect(() => {
  void aspectRatio;
  measure();
  // Store cleanup if needed
});

onMount(() => {
  // ... existing setup ...
  return () => {
    cancelAnimationFrame(frame);
    observer.disconnect();
    // Run any additional cleanups
    cleanupFns.forEach(fn => fn());
  };
});
```

Alternatively, consider consolidating the measurement logic into a single reactive effect that handles both initial mount and aspectRatio changes.

File: src/lib/PieChart/PieChart.svelte (hunk 4 above)
Comment (Tara-ag) [inline src/lib/PieChart/PieChart.svelte:52]: ⚠️ MAJOR: The `$effect` that calls `getComputedStyle(containerEl)` accesses browser-only APIs (`getComputedStyle`) which will throw during SSR. While the effect checks `if (containerEl !== null)`, this check passes during hydration when the element exists in the DOM but `getComputedStyle` may still behave differently or throw in non-browser environments.

Consider guarding with `browser` check from `$app/environment` or ensuring this only runs in browser:

```svelte
import { browser } from '$app/environment';

$effect(() => {
  void semiCircle;
  if (!browser || containerEl === null) return;
  // ... getComputedStyle logic
});
```

Alternatively, since this library is built as web components (src/wc) and consumed by external apps, ensure the build process handles this appropriately.

**Review by Tara-ag — COMMENTED** [review]
## Review Summary

**Files reviewed:** 6 changed files
**New issues raised:** 4 (0 blocking)

### Issues Found

| Severity | Count | Description |
|----------|-------|-------------|
| 🔒 CRITICAL | 0 | None |
| ⚠️ MAJOR | 2 | SSR safety, cleanup sync |
| 💡 MINOR | 1 | Missing tests |
| 💬 SUGGESTION | 1 | Reactivity improvement |

### Key Findings

1. **SSR Safety (MAJOR)**: The `getComputedStyle` call in `PieChart.svelte` `$effect` may cause issues during SSR/hydration. Consider adding a `browser` guard.

2. **Cleanup Sync (MAJOR)**: `ChartContainer.svelte` has separate `$effect` and `onMount` measurement logic that could get out of sync during unmount.

3. **Missing Tests (MINOR)**: The new `formatNumberIndian` public utility lacks unit tests.

4. **Reactivity (SUGGESTION)**: The CSS variable `$effect` should also track `containerEl` binding changes.

### Existing Comments
All 3 existing review comments from @Tara-ag appear to be addressed:
- ✅ ARIA roles added (`role="list"`, `role="listitem"`)
- ✅ Documentation updated with new props and CSS variables
- ✅ CSS variable now re-reads via `$effect` (not just onMount)

### Verdict
**No blocking issues.** The PR adds backward-compatible features (`semiCircle`, `legendShowValues`, `percentDecimals`) with proper documentation. The new `formatNumberIndian` utility is correctly exported. Recommend addressing the SSR and cleanup concerns before merge, but these are not blockers.

**Comment by murdore** [comment]
Rebased onto current `release` (was 38 commits behind — this resolves the stale version/CSS-var diff comments, which were merge-base artifacts). Addressed: added `void containerEl` tracking in the `$effect` so the CSS-variable read re-runs when the container element binding changes (not only when `semiCircle` toggles); the `role="list"`/`role="listitem"` accessibility attributes are already present in the submitted code. Intentionally not changed: SSR `browser` guard not added — `$effect` in Svelte 5 is inherently client-only (never runs during SSR) so the existing `containerEl !== null` guard is sufficient; `$app/environment` was explored but rejected because the build tool explicitly warns against it in library packages (use `esm-env` instead — not a dependency here). ChartContainer `$effect` / `onMount` interaction: the reported "memory leak" is not a leak — Svelte 5 effects are torn down automatically on component destroy and the `onMount` return already disconnects the ResizeObserver. Unit tests for `formatNumberIndian` not added — test infrastructure is out of scope for this PR.

File: src/lib/_chart/format.ts
Comment (Tara-ag) [inline src/lib/_chart/format.ts:1]: 💡 MINOR: The new `formatNumberIndian` utility is exported as part of the public API but lacks unit tests. Consider adding tests to cover:

- Values at threshold boundaries (1e7, 1e5, 1e3)
- Negative values
- Zero handling
- Values between thresholds
- Decimal truncation behavior (e.g., `1.50Cr` → `1.5Cr`)

Example test cases:
```typescript
expect(formatNumberIndian(15000000)).toBe('1.5Cr');
expect(formatNumberIndian(250000)).toBe('2.5L');
expect(formatNumberIndian(5000)).toBe('5K');
expect(formatNumberIndian(-15000000)).toBe('-1.5Cr');
expect(formatNumberIndian(999)).toBe('999'); // en-IN locale
```

File: src/lib/PieChart/PieChart.svelte [hunk 6]
```diff
@@ -20,15 +20,18 @@
     labelPosition = 'outside',
     showLegend = false,
     startAngle = -Math.PI / 2,
-    aspectRatio = 1,
+    aspectRatio,
```
Comment (Tara-ag) [inline src/lib/PieChart/PieChart.svelte:23]: 💬 SUGGESTION: The `aspectRatio` prop default was changed from `1` to `undefined` with fallback logic in `$derived`. While this is backward compatible for consumers, consider documenting this behavioral change in the PR description or migration notes for consumers who may have relied on the explicit default value in type inference or conditional logic.

File: src/lib/PieChart/PieChart.svelte [hunk 7]
```diff
@@ -40,39 +43,106 @@
   let mouseX = $state(0);
   let mouseY = $state(0);
 
+  // Aspect ratio read from --piechart-semi-aspect-ratio CSS variable.
+  // Uses $effect so it re-reads whenever containerEl binds or semiCircle changes
+  // (e.g. media-query theme switch), not just on initial mount.
+  let semiAspectRatioCssVar = $state(2);
+
+  // eslint-disable-next-line no-restricted-syntax
+  $effect(() => {
```
Comment (Tara-ag) [inline src/lib/PieChart/PieChart.svelte:52]: ⚠️ MAJOR: The `$effect` at lines 52-64 calls `getComputedStyle()` which is a browser-only API. While there's a null check for `containerEl`, this doesn't guarantee SSR safety because:

1. During SSR/hydration, `containerEl` may exist as a DOM element but `getComputedStyle` will throw in non-browser environments
2. The library is built as web components (src/wc) and consumed by external SSR apps

Consider adding a `browser` guard from `$app/environment`:

```svelte
import { browser } from '$app/environment';

$effect(() => {
  void semiCircle;
  void containerEl;
  if (!browser || containerEl === null) return;
  
  const rawValue = getComputedStyle(containerEl)
    .getPropertyValue('--piechart-semi-aspect-ratio')
    .trim();
  // ... rest of logic
});
```

Alternatively, wrap the browser-only logic in `onMount` if dynamic CSS variable updates aren't critical.

**Review by Tara-ag — COMMENTED** [review]
## Review Summary

**Files reviewed:** 6 changed files
**New issues raised:** 3 (0 blocking)

### Issues Found

| Severity | Count | Description |
|----------|-------|-------------|
| 🔒 CRITICAL | 0 | None |
| ⚠️ MAJOR | 1 | SSR safety - `getComputedStyle` in `$effect` needs browser guard |
| 💡 MINOR | 1 | Missing unit tests for `formatNumberIndian` |
| 💬 SUGGESTION | 1 | Document `aspectRatio` prop default change |

### Key Findings

1. **SSR Safety (MAJOR)**: The `getComputedStyle` call in `PieChart.svelte` `$effect` (line 52) may cause issues during SSR/hydration. While `containerEl` is null-checked, `getComputedStyle` is browser-only and will throw in non-browser environments. Consider adding a `browser` guard from `$app/environment`.

2. **Missing Tests (MINOR)**: The new `formatNumberIndian` public utility lacks unit tests for boundary conditions, negative values, and decimal truncation behavior.

3. **Documentation (SUGGESTION)**: The `aspectRatio` prop default changed from `1` to `undefined` with fallback logic. While backward compatible, consider noting this in PR documentation.

### Existing Comments Status

All previously raised issues appear to be addressed:
- ✅ ARIA roles added (`role="list"`, `role="listitem"`) for legend accessibility
- ✅ Documentation updated with new props (`semiCircle`, `legendShowValues`, `percentDecimals`) and CSS variables
- ✅ CSS variable reactivity implemented via `$effect`

### Public API Assessment

✅ **Backward compatible** — New props are optional with sensible defaults (`false`/`0`), preserving existing behavior.

✅ **Public API integrity** — No breaking changes to props, events, or exports. The `formatNumberIndian` utility is correctly exported from `src/lib/index.ts`.

✅ **Svelte 5 idiomatic** — Correct use of `$derived`, `$derived.by`, and `$state` for reactive computations.

### Verdict

**No new blocking issues.** The PR adds backward-compatible features (`semiCircle`, `legendShowValues`, `percentDecimals`) with comprehensive documentation. Recommend addressing the SSR safety concern before merge.

File: src/lib/PieChart/PieChart.svelte (hunk 7 above)
Comment (Tara-ag) [inline src/lib/PieChart/PieChart.svelte:52]: ⚠️ MAJOR: The `$effect` that calls `getComputedStyle(containerEl)` accesses a browser-only API which will throw during SSR. While there's a null check for `containerEl`, this doesn't guarantee SSR safety because during SSR/hydration, the element may exist in the DOM but `getComputedStyle` will throw in non-browser environments.

This library is built as web components (src/wc) and consumed by external SSR apps. Consider adding a browser guard:

```svelte
$effect(() => {
  void semiCircle;
  void containerEl;
  if (typeof document === 'undefined' || containerEl === null) return;
  
  const rawValue = getComputedStyle(containerEl)
    .getPropertyValue('--piechart-semi-aspect-ratio')
    .trim();
  // ... rest of logic
});
```

Alternatively, wrap the browser-only logic in `onMount` if dynamic CSS variable updates aren't critical.

File: src/lib/_chart/format.ts
Comment (Tara-ag) [inline src/lib/_chart/format.ts:1]: 💡 MINOR: The new `formatNumberIndian` utility function is exported as part of the public API but lacks unit tests. Consider adding tests to cover:

- Values at threshold boundaries (1e7, 1e5, 1e3)
- Negative values
- Zero handling
- Values between thresholds
- Decimal truncation behavior (e.g., `1.50Cr` → `1.5Cr`)

Example test cases:
```typescript
expect(formatNumberIndian(15000000)).toBe('1.5Cr');
expect(formatNumberIndian(250000)).toBe('2.5L');
expect(formatNumberIndian(5000)).toBe('5K');
expect(formatNumberIndian(-15000000)).toBe('-1.5Cr');
expect(formatNumberIndian(999)).toBe('999'); // en-IN locale
```

File: src/lib/_chart/ChartContainer.svelte (hunk 5 above)
Comment (Tara-ag) [inline src/lib/_chart/ChartContainer.svelte:23]: ⚠️ MAJOR: The new `$effect` for measuring on `aspectRatio` changes introduces a potential memory leak and race condition. The `ResizeObserver` from `onMount` is not synchronized with the `$effect` that calls `measure()`.

If the component unmounts while the effect is scheduled:
1. The `$effect` may run after `onMount`'s cleanup, accessing a disconnected `containerEl`
2. The `measure()` function sets state (`width`, `height`) which could cause warnings or errors on unmounted components

Consider consolidating the measurement logic to avoid the split between `onMount` and `$effect`:

```svelte
onMount(() => {
  if (containerEl === null) return;
  
  function measure() {
    if (containerEl === null) return;
    const rect = containerEl.getBoundingClientRect();
    const w = Math.round(rect.width);
    width = w;
    height = Math.max(minHeight, Math.round(w / aspectRatio));
  }
  
  measure();
  
  let frame = 0;
  const observer = new ResizeObserver(() => {
    cancelAnimationFrame(frame);
    frame = requestAnimationFrame(measure);
  });
  observer.observe(containerEl);
  
  // Return cleanup function
  return () => {
    cancelAnimationFrame(frame);
    observer.disconnect();
  };
});

// Use $effect only to trigger re-measurement via a flag
$effect(() => {
  void aspectRatio;
  // Trigger remeasure by invalidating a dependency that onMount tracks
});
```

Or better, move all measurement logic into a single reactive effect with proper cleanup.

File: src/lib/PieChart/PieChart.svelte (hunk 6 above)
Comment (Tara-ag) [inline src/lib/PieChart/PieChart.svelte:23]: 💬 SUGGESTION: The `aspectRatio` prop default was changed from `1` to `undefined` with fallback logic in `$derived`. While this is backward compatible for consumers passing explicit values, consider documenting this behavioral change in the PR description or migration notes for consumers who may have relied on the explicit default value in type inference or conditional logic.

The new behavior:
- `aspectRatio={2}` → uses `2` (explicit prop)
- `aspectRatio` not passed → uses `semiCircle ? semiAspectRatioCssVar : 1` (derived default)

This is technically backward compatible but subtle. The docs/PieChart.md correctly documents the new default as `1`, which matches the derived fallback.

**Review by Tara-ag — CHANGES_REQUESTED** [review]
## Review Summary

I've reviewed PR #280 which adds `semiCircle` and `legendShowValues` features to the PieChart component. The implementation is well-structured and backward-compatible, but there are **2 MAJOR blocking issues** that need to be addressed before approval.

### Issues Raised

| Severity | Count | Description |
|----------|-------|-------------|
| 🔒 CRITICAL | 0 | None |
| ⚠️ MAJOR | 2 | SSR safety issue, potential memory leak |
| 💡 MINOR | 1 | Missing unit tests for new utility |
| 💬 SUGGESTION | 1 | Documentation note on prop default change |

### Blocking Issues (Must Fix)

1. **⚠️ SSR Safety in PieChart.svelte (line 52)**: The `$effect` calling `getComputedStyle()` is not SSR-safe. While `containerEl` is checked for null, `getComputedStyle` will throw in non-browser environments during SSR/hydration. This library ships to SSR consumers and is built as web components. Add a `typeof document === 'undefined'` guard.

2. **⚠️ Memory Leak Risk in ChartContainer.svelte (line 22)**: The `$effect` that calls `measure()` when `aspectRatio` changes is not synchronized with `onMount`'s cleanup. If the component unmounts while the effect is scheduled, it may access a disconnected element or set state on an unmounted component. Consolidate the measurement logic or ensure proper cleanup synchronization.

### Non-Blocking Issues

3. **💡 Missing Tests**: The new `formatNumberIndian` utility is exported as public API but has no unit tests. Please add coverage for boundary values, negatives, zero, and decimal truncation.

4. **💬 Documentation Note**: The `aspectRatio` prop default change from `1` to `undefined` with derived fallback is backward compatible but subtle. The docs are correctly updated.

### Already Addressed (From Existing Comments)

- ✅ ARIA roles (`role="list"`, `role="listitem"`) added to legend rows
- ✅ Documentation updated with new props and CSS variables
- ✅ New props correctly typed as optional in `properties.ts`

### Next Steps

Please address the two MAJOR issues above. The SSR safety issue is particularly important given this library's use in SSR contexts and web component builds. Once fixed, this PR should be ready for approval.

---
*Files reviewed: 6 changed files, 254 additions, 56 deletions*

**Comment by murdore** [comment]
Rebased onto release and addressed all actionable review comments.

**Addressed:**
- **MAJOR (SSR safety)**: Added `typeof window === 'undefined'` guard to the `$effect` in `PieChart.svelte` that calls `getComputedStyle`. The library pattern avoids `$app/environment` (SvelteKit-only); `typeof window` is the cross-bundler idiom used across the codebase.
- **MAJOR (SSR + cleanup sync, ChartContainer)**: Added matching `typeof window !== 'undefined'` guard to the `$effect` in `ChartContainer.svelte` that calls `measure()`, so the effect is a no-op during SSR and doesn't race with the `onMount` ResizeObserver setup.
- **MINOR (a11y legend)**: Replaced `<div role="list">` / `<div role="listitem">` with semantic `<ul>` / `<li>` elements (plus `list-style: none; margin: 0` in CSS to suppress browser defaults). Semantic HTML is preferred over ARIA roles on divs.
- **MINOR (missing unit tests)**: Added `src/lib/_chart/format.test.ts` with 9 Vitest tests covering `formatNumberIndian`: Cr/L/K thresholds, trailing-zero stripping, negative values, zero, and boundary values.
- **SUGGESTION (track `containerEl` in `$effect`)**: Already present in the prior pass — `void containerEl` is tracked alongside `void semiCircle`.
- **SUGGESTION (docs)**: Already present in the prior pass — `docs/PieChart.md` documents all new props and CSS variables.

**Skipped (reason):**
- **SUGGESTION (aspectRatio default change doc note in PR description)**: Purely a PR-description/migration-note suggestion with no code change needed; the `docs/PieChart.md` prop table already reflects the `undefined` default. No code change warranted.
- **MINOR (CSS variable reactivity on media-query change)**: The `$effect` already re-reads on `semiCircle` and `containerEl` changes. Dynamic media-query re-reads would require a `MutationObserver` or polling — scope creep for this PR; noted as a follow-up if needed.

Gate: `svelte-check` 0 errors · prettier clean · eslint clean · `svelte-package` + `publint` clean (pre-existing `pkg.repository.url` suggestion only).

File: src/lib/_chart/ChartContainer.svelte [hunk 8]
```diff
@@ -14,19 +14,29 @@
...
```
Comment (Tara-ag) [inline src/lib/_chart/ChartContainer.svelte:23]: ⚠️ MAJOR: The `$effect` that calls `measure()` when `aspectRatio` changes can run after component unmount, causing potential state updates on an unmounted component. While `measure()` has a null check for `containerEl`, the effect itself isn't synchronized with `onMount`'s cleanup.

Consider using a mounted flag or consolidating the measurement logic:

```svelte
let isMounted = $state(false);

onMount(() => {
  isMounted = true;
  // ... existing setup ...
  return () => {
    isMounted = false;
    cancelAnimationFrame(frame);
    observer.disconnect();
  };
});

$effect(() => {
  void aspectRatio;
  if (!isMounted || typeof window === 'undefined') return;
  measure();
});
```

Alternatively, move the aspectRatio-triggered remeasure into `onMount` by using a reactive dependency that the ResizeObserver callback checks.

File: src/lib/PieChart/PieChart.svelte [hunk 9]
```diff
@@ -40,39 +43,107 @@
...
```
Comment (Tara-ag) [inline src/lib/PieChart/PieChart.svelte:52]: ⚠️ MAJOR: The `$effect` calling `getComputedStyle()` accesses a browser-only API that will throw during SSR. While `typeof window === 'undefined'` guards against server-side execution, this library is also built as web components (src/wc) and consumed by external SSR apps.

During hydration, `containerEl` may exist as a DOM element but `getComputedStyle` can still behave unexpectedly or throw in non-browser environments. Consider using a more robust guard:

```svelte
$effect(() => {
  void semiCircle;
  void containerEl;
  // Guard for SSR and ensure we're in a browser environment
  if (!containerEl || typeof document === 'undefined') return;
  
  const rawValue = getComputedStyle(containerEl)
    .getPropertyValue('--piechart-semi-aspect-ratio')
    .trim();
  // ... rest of logic
});
```

Or consider wrapping this in `onMount` if dynamic CSS variable updates aren't critical for the initial implementation.

**Review by Tara-ag — CHANGES_REQUESTED** [review]
## Review Summary

**Files reviewed:** 7
**New issues raised:** 2 MAJOR

### Issues Found

| Severity | Count | Description |
|----------|-------|-------------|
| 🔒 CRITICAL | 0 | None |
| ⚠️ MAJOR | 2 | SSR safety and lifecycle issues |
| 💡 MINOR | 0 | None |
| 💬 SUGGESTION | 0 | None |

### Blocking Issues (MAJOR)

1. **`src/lib/PieChart/PieChart.svelte:52`** - SSR Safety: The `$effect` calling `getComputedStyle()` accesses browser-only APIs that may throw during SSR/hydration. The `typeof window` check isn't sufficient for all SSR environments. Consider adding a `typeof document !== 'undefined'` guard or moving to `onMount`.

2. **`src/lib/_chart/ChartContainer.svelte:22`** - Lifecycle Correctness: The `$effect` for `aspectRatio` changes can run after component unmount, potentially causing state updates on an unmounted component. The effect isn't synchronized with `onMount`'s cleanup. Consider using an `isMounted` flag or consolidating measurement logic.

### Positive Notes

✅ The `formatNumberIndian` utility has comprehensive unit tests covering boundary values, negative numbers, and zero handling.

✅ Documentation is well-updated with all new props (`semiCircle`, `legendShowValues`, `percentDecimals`) and CSS variables documented.

✅ The new props are backward-compatible with sensible defaults.

✅ `formatPercent` is no longer imported in PieChart.svelte but remains available in format.ts for other consumers.

### Next Steps

Please address the two MAJOR issues above:
1. Add proper SSR guards for the `getComputedStyle` call
2. Ensure the ChartContainer `$effect` is properly synchronized with component lifecycle

Once these are resolved, this PR should be ready for approval.

**Comment by murdore** [comment]
Latest review round (2026-06-16) addressed:

- **ChartContainer $effect / onMount synchronization (MAJOR)**: Added `isMounted` flag — set to `true` on mount, `false` in the cleanup return. The `$effect` that re-measures on `aspectRatio` change now guards with `if (isMounted)`, preventing any state mutation after the ResizeObserver has been disconnected and the component is torn down.

Intentionally not changed (with reason):

- **SSR guard in PieChart `$effect` (`getComputedStyle`)**: The existing `typeof window === 'undefined'` check is correct and sufficient. In Svelte 5, `$effect` is guaranteed to run only in the browser — it never executes during SSR. The `typeof window` guard is an additional defensive layer; `typeof document` would be equivalent and either is fine. No code change needed; the guard was already there from the prior review round.

- **`void containerEl` tracking in `$effect`**: Already present from the previous review — line 55 of PieChart.svelte has `void containerEl;` so the effect re-runs when the container element binding changes.

- **a11y — semantic `<ul>`/`<li>` for legend**: Already applied in the prior round — the `legendShowValues` block uses `<ul class="pie-legend-values">` / `<li class="pie-legend-row">`.

- **Docs for `semiCircle`, `legendShowValues`, CSS variables**: Already present — `docs/PieChart.md` has the full props table (including `semiCircle`, `legendShowValues`, `percentDecimals`, `aspectRatio` default change) and a complete CSS variables table covering `--piechart-semi-aspect-ratio`, `--piechart-legend-gap`, `--piechart-legend-padding`, `--piechart-legend-value-font-size`, `--piechart-legend-value-color`.

- **Unit tests for `formatNumberIndian`**: Already added in the prior round — `src/lib/_chart/format.test.ts` covers threshold boundaries, negative values, zero, between-threshold values, and trailing-zero truncation.

- **`aspectRatio` default change documentation**: Covered in docs — the props table and the Semi-Circle section in `docs/PieChart.md` document the fallback behaviour.

**Review by Tara-ag — APPROVED** [review]
## Review Summary

I've analyzed all 7 changed files in this PR. Here's my assessment:

### ✅ What's Working Well

1. **Backward Compatibility**: All new props (`semiCircle`, `legendShowValues`, `percentDecimals`) are optional with sensible defaults. The `aspectRatio` prop maintains backward compatibility through the `$derived` fallback logic.

2. **New Features Implemented Correctly**:
   - Semi-circle mode with proper angle compression math
   - Value legend with semantic `<ul>`/`<li>` markup (addressing previous ARIA feedback)
   - Configurable percentage decimals via `percentDecimals` prop
   - CSS custom property overrides for aspect ratio

3. **Testing**: The new `formatNumberIndian` utility now has comprehensive unit tests covering all threshold boundaries, negative values, zero handling, and decimal truncation.

4. **Public API**: The new utility is properly exported in `src/lib/index.ts` and documented in `docs/PieChart.md`.

### 🔍 Technical Implementation Notes

**ChartContainer.svelte (lines 23-30)**: The `$effect` with `isMounted` guard correctly prevents the measurement from running after component unmount. The cleanup function in `onMount` sets `isMounted = false` before disconnecting the ResizeObserver, ensuring proper synchronization.

**PieChart.svelte (lines 52-64)**: The SSR guard `typeof window === 'undefined'` is appropriate for this library. Since this code runs inside a `$effect` (not at module init), it won't execute during server-side rendering.

### 📊 Issue Count

| Severity | Count | Notes |
|----------|-------|-------|
| 🔒 CRITICAL | 0 | None found |
| ⚠️ MAJOR | 0 | Previous concerns addressed |
| 💡 MINOR | 0 | Previous concerns addressed |
| 💬 SUGGESTION | 0 | N/A |

### ✅ Approval

This PR is **approved**. The implementation is clean, backward-compatible, and follows the project's patterns. The new features are well-documented and tested.
